src/commands/run.ts

```diff
@@ -405,18 +406,56 @@ const fluenceRun = async (args: RunArgs) => {
   }
 
   await initFluenceClient(args, args.maybeFluenceConfig);
-  const { Fluence, callAquaFunction } = await import("@fluencelabs/js-client");
+
+  const { Fluence, callAquaFunction, js2aqua, aqua2js } = await import(
+    "@fluencelabs/js-client"
+  );
+
+  const schema = functionCall.funcDef;
+
+  // TODO: remove this after DXJ-535 is done
+  const validatedRunData = Object.fromEntries(
+    Object.entries(args.runData ?? {}).map(([argName, argValue]) => {
+      assert(
+        typeof argValue !== "function",
+        "Should be impossible to pass function as an argument to fluence run",
+      );
+
+      const fields =
+        schema.arrow.domain.tag === "nil" ? {} : schema.arrow.domain.fields;
+
+      const argSchema = fields[argName];
+
+      assert(
+        argSchema !== undefined,
+        "Should be impossible because schema always contains same keys as runData",
+      );
+
+      assert(
+        argSchema.tag !== "arrow",
+        "Should be impossible to pass function as an argument to fluence run",
+      );
+
+      return [argName, js2aqua(argValue, argSchema, { path: [argName] })];
+    }),
+  );
 
   const result = await callAquaFunction({
-    args: args.runData ?? {},
+    script: functionCall.script,
     config: {},
-    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
-    def: functionCall.funcDef,
     peer: Fluence.getClient(),
-    script: functionCall.script,
+    args: validatedRunData,
   });
 
-  return result;
+  const returnSchema =
+    schema.arrow.codomain.tag === "unlabeledProduct" &&
+    schema.arrow.codomain.items.length === 1
+      ? schema.arrow.codomain.items[0]
+      : schema.arrow.codomain;
+
+  assert(returnSchema !== undefined, "This value cannot be 'undefined'");
```

Comment:
if it's impossible, better explain why

Reply:
undefined appears in union after making checked indexed access to an array `items`.
Length of this has been checked already.
To me it looks ridiculous to put an assert after every array or object access.
Useless runtime work.

Reply:
Updated description there